### ISSUE TITLE
/v2/user-tasks/search throws Exception

### ISSUE THREAD

**Issue by kristofferjalen** [camunda/camunda#24146]
**Describe the bug**

Running the new “Query user tasks” operation in Camunda 8 API throws an Exception. Yes, I have enabled the alpha feature.

See:

- https://forum.camunda.io/t/query-user-tasks-alpha-throws-error/56419
- https://forum.camunda.io/t/camunda-8-api-unknown-sortby/56412

**To Reproduce**

```
POST http://localhost:8088/v2/user-tasks/search
Content-Type: application/json
Authorization: Bearer {{access_token}}

{
  "filter": {
    "processInstanceKey": 1234
  }
}
```

**Expected behavior**

A response message with HTTP status code 200.

**Actual behavior**

```
{
  "type": "about:blank",
  "title": "Failed to execute UserTask Search Query",
  "status": 500,
  "detail": "Failed to execute search query",
  "instance": "/v2/user-tasks/search"
}
```

**Environment:**
Latest docker-compose file.

**Comment by megglos** [comment]
Hey @kristofferjalen ,

thank you for reaching out! Could you also provide the logs from the C8 gateway especially on the case where the error is rather generic `Failed to execute search query` this would be useful for us. We will try to reproduce it regardless though.

@tmetzke can you please take a look here or delegate? At least in one case the error indicated something is off with the sortBy see https://forum.camunda.io/t/camunda-8-api-unknown-sortby/56412. 

Best
Sebastian

**Comment by mihail-ca** [comment]
@tmetzke I was only able to reproduce the 500 Internal Server error when the index for User Tasks does not exist. In this case, ES throws index not found error, which in turn we return as 500 Internal Server error to the customer. But there are at least two issues here:
- currently the v2 REST API does not log any errors, not even at DEBUG level, so we cannot know for sure the customer setup / error, it could have been something else. I also raised this in the [document here](https://docs.google.com/document/d/1n31TbduE2WyRqme-8jssaqpoTVA2jEXAvvNKY0H2RW8/edit?tab=t.0#heading=h.bc9qluclaws), but maybe it's the wrong place 
- for the case the index does not exist, should we return 500 Internal Server error, or 404 Not Found. Because the new Zeebe exporter, unlike the previous Operate / Tasklist implementations, only creates ES indices when it actually has something to write to them (not sure if by design). Operate for example, would always create all indices from the start.

As for the [sortBy error](https://forum.camunda.io/t/camunda-8-api-unknown-sortby/56412), this is expected, as we only have the following sort fields for User Tasks: https://github.com/camunda/camunda/blob/main/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/SearchQueryRequestMapper.java#L685

**Comment by tmetzke** [comment]
Thanks, @mihail-ca, super helpful insights 👍 

1. I agree that error logging must be improved, can you create an issue for this on our API board? That leads to better traceability and you can announce it in our Slack channel as well 👍 
2. Can you bring up the non-existing index in the exporter team channel and ask if that is intentional? I wasn't aware and I am not sure this was announced. This makes the logic in consumers more complex because you have to account for the missing index and figure out if it's fine (not created yet) or not (you have something misconfigured). Ideally, we don't have to handle this in the API search.
3. For the `sortBy`, we should fix this by documenting which fields are supported in the API spec and thus reference docs. Can you create an issue for this as well?

If you need any assistance with any of those topics, please don't hesitate to reach out, happy to help!

**Comment by mihail-ca** [comment]
> Thanks, @mihail-ca, super helpful insights 👍
> 
> 1. I agree that error logging must be improved, can you create an issue for this on our API board? That leads to better traceability and you can announce it in our Slack channel as well 👍
> 2. Can you bring up the non-existing index in the exporter team channel and ask if that is intentional? I wasn't aware and I am not sure this was announced. This makes the logic in consumers more complex because you have to account for the missing index and figure out if it's fine (not created yet) or not (you have something misconfigured). Ideally, we don't have to handle this in the API search.
> 3. For the `sortBy`, we should fix this by documenting which fields are supported in the API spec and thus reference docs. Can you create an issue for this as well?
> 
> If you need any assistance with any of those topics, please don't hesitate to reach out, happy to help!

@tmetzke :
1. Logging issue created here: https://github.com/camunda/camunda/issues/24332
2. Raised the issue in exporter channel here: https://camunda.slack.com/archives/C06F0GLJNFM/p1730742416718829
3. Not sure here, where to create the issue. Still in out C8 API board ?

Thank you

**Comment by tmetzke** [comment]
> 3. Not sure here, where to create the documentation issue. Still in our C8 API board ?

Yes, our board as well, thank you! 👍

**Comment by mihail-ca** [comment]
FYI created documentation issue for sort fields: https://github.com/camunda/camunda/issues/24363

**Comment by tmetzke** [comment]
**Update**

* The root cause for the 500 response is most probably a missing Elastic index that is currently only lazily created when there is something to write to it. If you use the endpoint beforehand, it will currently lead to 500. This will be fixed with https://github.com/camunda/camunda/issues/24412.
* The `sortBy` issue will be handled with https://github.com/camunda/camunda/issues/24363.

**Comment by tmetzke** [comment]
@mihail-ca, the root cause should be fixed with https://github.com/camunda/camunda/issues/24412. Can you recheck if this works from the REST API perspective as expected?
If this works, we can close this issue as done.
The https://github.com/camunda/camunda/issues/24363 follow-up is out of scope here and is a separate topic. Plus, that one is an improvement, not a bug.

**Comment by mihail-ca** [comment]
@tmetzke I still get the error, the index is not created eagerly by Exporter. But at least the error is more clear now:

```
{
  "type": "about:blank",
  "title": "co.elastic.clients.elasticsearch._types.ElasticsearchException",
  "status": 500,
  "detail": "Unexpected error occurred during the request processing: [es/search] failed: [index_not_found_exception] no such index [tasklist-task-8.5.0_]",
  "instance": "/v2/user-tasks/search"
}
```

**Comment by korthout** [comment]
Reported in `8.7.0-alpha2-rc1` on SaaS INT [here](https://console.cloud.google.com/errors/detail/CODGq6mty7OVyQE;service=zeebe;time=P7D;locations=global?project=camunda-saas-int)

<details><summary>Stacktrace</summary>
<p>

<pre>
 co.elastic.clients.elasticsearch._types.ElasticsearchException: [es/search] failed: [index_not_found_exception] no such index [operate-process-8.3.0_alias]
	at co.elastic.clients.transport.ElasticsearchTransportBase.getApiResponse(ElasticsearchTransportBase.java:345) ~[elasticsearch-java-8.13.4.jar:?]
	at co.elastic.clients.transport.ElasticsearchTransportBase.performRequest(ElasticsearchTransportBase.java:147) ~[elasticsearch-java-8.13.4.jar:?]
	at co.elastic.clients.elasticsearch.ElasticsearchClient.search(ElasticsearchClient.java:1923) ~[elasticsearch-java-8.13.4.jar:?]
	at io.camunda.search.es.clients.ElasticsearchSearchClient.search(ElasticsearchSearchClient.java:55) ~[camunda-search-client-elasticsearch-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.search.clients.SearchClientBasedQueryExecutor.lambda$search$0(SearchClientBasedQueryExecutor.java:50) ~[camunda-search-client-query-transformer-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.search.clients.SearchClientBasedQueryExecutor.executeSearch(SearchClientBasedQueryExecutor.java:73) ~[camunda-search-client-query-transformer-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.search.clients.SearchClientBasedQueryExecutor.search(SearchClientBasedQueryExecutor.java:49) ~[camunda-search-client-query-transformer-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.search.clients.SearchClients.searchProcessDefinitions(SearchClients.java:206) ~[camunda-search-client-query-transformer-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.service.ProcessDefinitionServices.search(ProcessDefinitionServices.java:45) ~[camunda-service-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.zeebe.gateway.rest.controller.ProcessDefinitionQueryController.search(ProcessDefinitionQueryController.java:62) ~[zeebe-gateway-rest-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.zeebe.util.Either$Right.fold(Either.java:445) ~[zeebe-util-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at io.camunda.zeebe.gateway.rest.controller.ProcessDefinitionQueryController.searchProcessDefinitions(ProcessDefinitionQueryController.java:53) ~[zeebe-gateway-rest-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255) ~[spring-web-6.1.14.jar:6.1.14]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188) ~[spring-web-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:926) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:831) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:914) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:547) ~[jakarta.servlet-api-6.0.0.jar:6.0.0]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.1.14.jar:6.1.14]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:614) ~[jakarta.servlet-api-6.0.0.jar:6.0.0]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-embed-websocket-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:108) ~[spring-web-6.1.14.jar:6.1.14]
	at org.springframework.web.filter.CompositeFilter.doFilter(CompositeFilter.java:74) ~[spring-web-6.1.14.jar:6.1.14]
	at io.camunda.zeebe.gateway.RestApiCompositeFilter.doFilter(RestApiCompositeFilter.java:36) ~[camunda-zeebe-8.7.0-alpha2-rc1.jar:8.7.0-alpha2-rc1]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.1.14.jar:6.1.14]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.14.jar:6.1.14]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.1.14.jar:6.1.14]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.14.jar:6.1.14]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:113) ~[spring-web-6.1.14.jar:6.1.14]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.14.jar:6.1.14]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.1.14.jar:6.1.14]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.14.jar:6.1.14]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:731) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:384) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:905) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[tomcat-embed-core-10.1.30.jar:10.1.30]
	at java.base/java.lang.Thread.run(Unknown Source) [?:?] 
</pre>

</p>
</details>

**Comment by tmetzke** [comment]
For transparency, we believe it's a race condition: https://github.com/camunda/camunda/issues/24412#issuecomment-2545267657

**Comment by tmetzke** [comment]
Update: We'll discuss mechanisms to wait for schema creation before starting the REST API.
We'll update here as soon as we have news.

**Comment by tmetzke** [comment]
We'll fix this with https://github.com/camunda/camunda/issues/26890

**Comment by houssain-barouni** [comment]
Since 8.8.0-alpha4, the schema initialization is part of the application bootstrapping, so the API is only accessible when the schema is created/updated.
Closing this issue.
